plutus-tx/src/PlutusTx/Code.hs

```diff
@@ -55,11 +55,14 @@ data CompiledCodeIn uni fun a =
 -- | 'CompiledCodeIn' instantiated with default built-in types and functions.
 type CompiledCode = CompiledCodeIn PLC.DefaultUni PLC.DefaultFun
 
--- | Apply a compiled function to a compiled argument.
+-- | Apply a compiled function to a compiled argument. Will fail if the versions don't match.
 applyCode
     :: (PLC.Closed uni, uni `PLC.Everywhere` Flat, Flat fun)
-    => CompiledCodeIn uni fun (a -> b) -> CompiledCodeIn uni fun a -> CompiledCodeIn uni fun b
-applyCode fun arg = DeserializedCode (UPLC.applyProgram (getPlc fun) (getPlc arg)) (PIR.applyProgram <$> getPir fun <*> getPir arg) (getCovIdx fun <> getCovIdx arg)
+    => CompiledCodeIn uni fun (a -> b) -> CompiledCodeIn uni fun a -> Maybe (CompiledCodeIn uni fun b)
+applyCode fun arg = do
+  uplc <- UPLC.applyProgram (getPlc fun) (getPlc arg)
+  pir <- PIR.applyProgram <$> getPir fun <*> getPir arg
+  pure $ DeserializedCode uplc pir (getCovIdx fun <> getCovIdx arg)
```

Comment:
applyCode is often used with liftCode, so we should try to keep the same interface with liftCode / safeLiftCode.

So I suggest that applyCode remains with the old type signature:

```
applyCode :: code (a -> b) -> code a -> code b
applyCode = unsafely $ safeApplyCode
```

and we introduce the `safeApplyCode`:

```
safeApplyCode
    :: (PLC.Closed uni, uni `PLC.Everywhere` Flat, Flat fun, MonadError e m)
    => CompiledCodeIn uni fun (a -> b) -> CompiledCodeIn uni fun a -> m (CompiledCodeIn uni fun b)
safeApplyCode fun arg = do
  uplc <- maybe throwSth pure $ UPLC.applyProgram (getPlc fun) (getPlc arg)
  pir <- maybe throwSth pure $ PIR.applyProgram <$> getPir fun <*> getPir arg
  pure $ DeserializedCode uplc pir (getCovIdx fun <> getCovIdx arg)
```

Reply:
Oh wait, you mean just `applyCode`. Yes, that probably is worth having the extra function, I'll add it.

Reply:
I agree with @bezirg on keeping the same interface, though I like `unsafeFoo/foo` slightly better than `foo/safeFoo`. The former is much more common in Haskell.